docs: PDE Tutorial (#1091)
spectrocloud/librarium: content/docs/04.5-devx/06-apps/05-deploy-app.md

  

Once you have filled out all the required information, navigate to the top left side of the wizard screen and click on the **Actions** button **+**. Go ahead and select the **Container Deployment** to add the final service layer, the UI.

  

### UI

  

This time you will use a different container image for the UI that contains a reverse proxy. The reverse proxy is responsible for two important tasks. The first task is to route requests to the API inside the UI container instead of using your browser. The reason behind this behavior is to access the API server that is not publicly exposed. The reverse proxy will pick up the API request and forward it to the API container. The second task is to insert an authentication token in all requests to the API. The API server configuration you provided enabled authentication, so a Bearer token is required for all requests. The following diagram illustrates the network connectivity path and behavior discussed.

  
  

![A diagram of the reverse proxy architecture](/tutorials/deploy-app/devx_apps_deploys-apps_reverse-proxy-diagram.png)

  

> Go ahead and provide the UI container with the following information.

Author: ```suggestion
Provide the UI container with the following information.
```

      description: "The id of the virtual cluster that will be created with the resource spectrocloud_virtual_cluster.cluster-1.",
      tooltipPlacement: "rightTop",
    },
  ]}
>

```tf
resource "spectrocloud_application" "scenario-1" {
  name                    = "single-scenario"
  application_profile_uid = spectrocloud_application_profile.hello-universe-ui.id

  config {
    cluster_name      = spectrocloud_virtual_cluster.cluster-1.name
    cluster_uid = spectrocloud_virtual_cluster.cluster-1.id
  }
  tags = concat(var.tags, ["scenario-1"])
}
```

</PointsOfInterest>

You can preview the resources Terraform will create by issuing the following command.

```shell
terraform plan
```

```
// Output condensed for readability
Plan: 3 to add, 0 to change, 0 to destroy.

```

The output displays the resources Terraform will create in an actual implementation. If you review the output, you will find the three resources previously discussed in great detail.

Go ahead and deploy the application by using the `terraform apply` command.

```shell
terraform apply -auto-approve
```

```
// Output condensed for readability
Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

> Log in to [Palette](https://console.spectrocloud.com) and navigate to the left **Main Menu**, select **Apps**. Click on the **scenario-1** row, which will take you to the application’s overview page. Once you are on the scenario-1 overview page, click on the exposed URL for the service. A hyperlink for port 8080 is available.

Reviewer: ```suggestion
Log in to [Palette](https://console.spectrocloud.com), navigate to the left **Main Menu**, and select **Apps**. Click on the **scenario-1** row, which takes you to the application’s overview page. Once you are on the scenario-1 overview page, click on the exposed URL for the service. A hyperlink for port 8080 is available.
```

![A diagram of the reverse proxy architecture](devx_apps_deploys-apps_reverse-proxy-diagram.png)

To deploy the second scenario, you will again deploy the same three resource types previously discussed but another instance of them.

- `spectrocloud_virtual_cluster` - `cluster-2` - this resource will create the second virtual cluster. 

- `spectrocloud_application_profile` - `hello-universe-complete` - the application profile that will contain the three different services, database, API, and UI

> - `spectrocloud_application` - `scenario-2` - the application that will be deployed into cluster-2 that use the `spectrocloud_application_profile.hello-universe-complete` app profile.

Reviewer: ```suggestion
- `spectrocloud_application` - `scenario-2` - the application that will be deployed into cluster-2 that uses the `spectrocloud_application_profile.hello-universe-complete` app profile.
```

                secretName: "{{.spectro.system.appdeployment.tiername}}-custom-secret"
                ownerReference:
                  apiVersion: v1
                  kind: Service
                  name: "{{.spectro.system.appdeployment.tiername}}-svc"
                keyToCheck: spec.ports[0].port
        containerService:
            serviceName: "{{.spectro.system.appdeployment.tiername}}-svc"
            registryUrl: ""
            image: ${var.multiple_container_images["ui"]}
            access: public
            ports:
              - "8080"
              - "3000"
            env:
              - name: "API_URI"
                value: "http://{{.spectro.app.$appDeploymentName.api.CONTAINER_SVC_EXTERNALIP}}"
              - name: "SVC_URI"
                value: "http://{{.spectro.app.$appDeploymentName.api.CONTAINER_SVC}}:{{.spectro.app.$appDeploymentName.api.CONTAINER_SVC_PORT}}"
              - name: "TOKEN"
                value: "${var.token}"
            serviceType: LoadBalancer
    EOT   
  }
  tags = concat(var.tags, ["scenario-2"])
}
```

Each service has its own `pack {}` and a set of unique properties and values. 

> The database service block uses a different data resource, `data.spectrocloud_pack_simple.postgres_service`,  to find the Postgres service. If you review the data resource, you will find a different type, `operator-instance`. The Postgres service uses a Postgres [operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) to manage the database inside the virtual cluster. 

Reviewer: ```suggestion
The database service block uses a different data resource, `data.spectrocloud_pack_simple.postgres_service`, to find the Postgres service. If you review the data resource, you will find a different type, `operator-instance`. The Postgres service uses a Postgres [operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) to manage the database inside the virtual cluster. 
```

    values          = <<-EOT
        # ....
        # ....
        containerService:
            serviceName: "{{.spectro.system.appdeployment.tiername}}-svc"
            registryUrl: ""
            image: ${var.multiple_container_images["ui"]}
            access: public
            ports:
              - "8080"
              - "3000"
            env:
              - name: "API_URI"
                value: "http://{{.spectro.app.$appDeploymentName.api.CONTAINER_SVC_EXTERNALIP}}"
              - name: "SVC_URI"
                value: "http://{{.spectro.app.$appDeploymentName.api.CONTAINER_SVC}}:{{.spectro.app.$appDeploymentName.api.CONTAINER_SVC_PORT}}"
              - name: "TOKEN"
                value: "${var.token}"
            serviceType: LoadBalancer
    EOT   
  }

```

<InfoBox>

All container services expose their service address, Kubernetes hostname, and the exposed service ports as output variables.
You will use output variables frequently when creating app profiles in the future. You can learn more about connecting services by refering to the [Service Connectivity](/devx/app-profile/services/connectivity) documentation.  

</InfoBox>

> Open up the file **inputs.tf** and change the `variable enable-second-scenario"` and set its default value to `true`.

Reviewer: ```suggestion
Open the **inputs.tf** file and set the `variable enable-second-scenario"` default value to `true`.
```

<br />

```terraform
variable "enable-second-scenario" {
  type        = bool
  description = "Whether to enable the second scenario"
  default     = true
}
```

> Next, issue the command `terraform apply` to deploy the second scenario. Take notice how the `-var` flag is included with the token value in the command. 

Reviewer: ```suggestion
Next, issue the command `terraform apply` to deploy the second scenario. Notice how the `-var` flag is included with the token value in the command. 
```

## Clean-up

To remove all resources created in this tutorial, issue the `terraform destroy` command.

<br />

```shell
terraform destroy -var="token=931A3B02-8DCC-543F-A1B2-69423D1A0B94" -auto-approve
```

```
Destroy complete! Resources: 6 destroyed.
```

<br />

<InfoBox>

If a cluster remains in the delete phase for over 15 minutes, it becomes eligible for **Force Delete**. To trigger a force delete, navigate to the respective cluster’s details page and click on **Settings**. Click on the **Force Delete Cluster** to delete the cluster. Palette will automatically remove clusters stuck in the cluster deletion phase for over 24 hrs.

</InfoBox>

<br />

</Tabs.TabPane>

</Tabs>

# Wrap-Up

In this tutorial, you learned about Palette’s Dev Engine and App Mode. You deployed two virtual clusters, each containing a different architecture and configuration of the Hello Universe application. Palette’s Dev Engine enables developers to quickly deploy applications into a Kubernetes environment without requiring Kubernetes knowledge. In a matter of minutes, you deployed a new Kubernetes cluster and all its applications without having to write Kubernetes configuration files.

> To learn more about Palette Dev Engine and its capabilities check out the reference resource below.

Reviewer: ```suggestion
To learn more about Palette Dev Engine and its capabilities, check out the reference resource below.
```

This tutorial will teach you how to deploy single and multiple applications to Kubernetes through Palette’s Dev Engine experience. You will learn about *App Mode*, *App Profiles*, and *Palette Virtual Clusters* and understand how they enable you to deploy applications to Kubernetes quickly with minimal overhead. 

# Prerequisites

To complete this tutorial, you will need the following items.

- A Spectro Cloud account
- Basic knowledge about containers.

If you select the Terraform workflow, you will need the following software installed.
- Terraform v1.3.6 or greater
- Git v2.30.0 or greater 

There are no expenses associated with this tutorial as everything falls under the Palette Free Tier.

# Architecture 

The tutorial includes two scenarios, and for each scenario, you will deploy a separate Kubernetes environment. The following diagram illustrates the different layers that will power the tutorial environment.

> ![Architecutre diagram depicting two virtual clusters](/tutorials/deploy-app/devx_apps_deploy-apps_architecture-diagram.png)

Author: ```suggestion
![Architecture diagram depicting two virtual clusters](/tutorials/deploy-app/devx_apps_deploy-apps_architecture-diagram.png)
```

  

## Clean-up

  

To remove all resources created in this tutorial, begin by navigating to the left **Main Menu** and select **Apps**. For each application, click on the **three-dots Menu** to expand the options menu and click on the **Delete** button. Repeat this process for each application.

  

![Apps view with an arrow pointing towards the delete button](/tutorials/deploy-app/devx_apps_deploy-apps_delete-apps-view.png)

  

Next, in the left **Main Menu**, click on the **Cluster** link to access the clusters page.

Click on **cluster-1** to access its details page. Click on **Settings** from the details page to expand the settings menu. Click on **Delete** to delete the cluster. You will be asked to enter the cluster name to confirm the delete action. Type the cluster name to proceed with the delete step. Repeat this process for cluster-2.

  

![Delete a cluster view with arrow](/tutorials/deploy-app/devx_apps_deploy-apps_delete-cluster-view.png)

  

<InfoBox>
 

> If a cluster remains in the delete phase for over 15 minutes, it becomes eligible for **Force Delete**. To trigger a force delete, navigate to the respective cluster’s details page and click on **Settings**. Click on the **Force Delete Cluster** to delete the cluster. Palette will automatically remove clusters stuck in the cluster deletion phase for over 24 hrs.

Reviewer: ```suggestion
If a cluster remains in the delete phase for over 15 minutes, it becomes eligible for **Force Delete**. To trigger a force delete, navigate to the respective cluster’s details page and click on **Settings**. Click on the **Force Delete Cluster** to delete the cluster. Palette will automatically remove clusters stuck in the cluster deletion phase for over 24 hours.
```

<br />

```tf
resource "spectrocloud_virtual_cluster" "cluster-1" {
  name              = var.scenario-one-cluster-name
  cluster_group_uid = data.spectrocloud_cluster_group.beehive.id

  resources {
    max_cpu           = 4
    max_mem_in_mb     = 4096
    min_cpu           = 0
    min_mem_in_mb     = 0
    max_storage_in_gb = "2"
    min_storage_in_gb = "0"
  }

  tags = concat(var.tags, ["scenario-1"])

  timeouts {
    create = "15m"
    delete = "15m"
  }
}

```

The cluster group id is retrieved from the data resource `spectrocloud_cluster_group.beehive`. The data resource will query the Palette API and retrieve information about the specified cluster group, which is the *beehive* cluster group made available for all Palette users. This resource will create a new virtual cluster that is hosted in the cluster group, beehive.

<br />

```tf
data "spectrocloud_cluster_group" "beehive" {
  name    = var.cluster-group-name
  context = "system"
}
```

> Next, take a look at the the **application-profiles.tf** file. The resource `spectrocloud_application_profile.hello-universe-ui` is the resource responsible for creating the app profile for the first scenario. There are several points of interest in this resource that you should be familiar with. Focus on these five key points:

Author: ```suggestion
Next, take a look at the **application-profiles.tf** file. The resource `spectrocloud_application_profile.hello-universe-ui` is the resource responsible for creating the app profile for the first scenario. There are several points of interest in this resource that you should be familiar with. Focus on these five key points:
```

                ownerReference:
                  apiVersion: v1
                  kind: Service
                  name: "{{.spectro.system.appdeployment.tiername}}-svc"
                keyToCheck: spec.ports[0].port
        containerService:
            serviceName: "{{.spectro.system.appdeployment.tiername}}-svc"
            registryUrl: ""
            image: ${var.single-container-image}
            access: public
            ports:
              - "8080"
            serviceType: load balancer
    EOT
  }
  tags = concat(var.tags, ["scenario-1"])
}
```

</PointsOfInterest>

A tip for gathering the required values to provide the `values` attribute is to visit the Palette console and create the app profile through the UI. During the app profile creation process, click on the API button to display the API payload. Review the payload's `values` attribute to find all of the properties of the service. You can copy the entire string and pass it to the resource `spectrocloud_application_profile` as an input for the `values` attribute.

![UI's ability to display the API object](/tutorials/deploy-app/devx_apps_deploy-apps_ui-api-display.png)

> The last Terraform resource to review before deploying the application is located in the **application.tf** file. The resource `spectrocloud_application.hello-universe-ui` is what creates the *App*. In Palette, an app combines a virtual cluster and an app profile. When you deploy an app profile into a virtual cluster, you create an app. This resource points to the app profile `spectrocloud_application_profile.hello-universe-ui` and the cluster resource `spectrocloud_virtual_cluster.cluster-1`. The two resources are required to create an app.

Author: ```suggestion
The last Terraform resource to review before deploying the application is located in the **application.tf** file. The resource `spectrocloud_application.hello-universe-ui` is what creates the *app*. In Palette, an app combines a virtual cluster and an app profile. When you deploy an app profile into a virtual cluster, you create an app. This resource points to the app profile `spectrocloud_application_profile.hello-universe-ui` and the cluster resource `spectrocloud_virtual_cluster.cluster-1`. The two resources are required to create an app.
```

<br />

```
pack {
    name            = "api"
    type            = data.spectrocloud_pack_simple.container_pack.type
    registry_uid    = data.spectrocloud_registry.public_registry.id
    source_app_tier = data.spectrocloud_pack_simple.container_pack.id
    values          = <<-EOT
pack:
  namespace: "{{.spectro.system.appdeployment.tiername}}-ns"
  releaseNameOverride: "{{.spectro.system.appdeployment.tiername}}"
postReadinessHooks:
  outputParameters:
  #....
  #...
containerService:
  serviceName: "{{.spectro.system.appdeployment.tiername}}-svc"
  registryUrl: ""
  image: ${var.multiple_container_images["api"]}
  access: public
  ports:
    - "3000"
  serviceType: load balancer
  env:
    - name: DB_HOST
      value: "{{.spectro.app.$appDeploymentName.postgres-db.POSTGRESMSTR_SVC}}"
    - name: DB_USER
      value: "{{.spectro.app.$appDeploymentName.postgres-db.USERNAME}}"
    - name: DB_PASSWORD
      value: "{{.spectro.app.$appDeploymentName.postgres-db.PASSWORD}}"
    - name: DB_NAME
      value: counter
    - name: DB_INIT
      value: "true"
    - name: DB_ENCRYPTION
      value: "${var.database-ssl-mode}"
    - name: AUTHORIZATION
      value: "true"
    EOT
  }
```

> The last `pack {}` block in the app profile resource `spectrocloud_application_profile.hello-universe-complete` is for the UI. Like the API service, environment variables are used to initialize UI and the reverse proxy. The UI service requires the URL of the API service and the URL of the public-facing load balancer. Palette output variables are used to populate these two environment variables. A Terraform variable will populate the authentication token required for all API requests.

Reviewer: ```suggestion
The last `pack {}` block in the app profile resource `spectrocloud_application_profile.hello-universe-complete` is for the UI. Like the API service, environment variables are used to initialize the UI and the reverse proxy. The UI service requires the URL of the API service and the URL of the public-facing load balancer. Palette output variables are used to populate these two environment variables. A Terraform variable will populate the authentication token required for all API requests.
```

              - name: "TOKEN"
                value: "${var.token}"
            serviceType: load balancer
    EOT   
  }

```

<InfoBox>

All container services expose their service address, Kubernetes hostname, and the exposed service ports as output variables.
You will use output variables frequently when creating app profiles in the future. You can learn more about connecting services by referring to the [Service Connectivity](/devx/app-profile/services/connectivity) documentation.  

</InfoBox>

Open the **inputs.tf** file and set the `variable enable-second-scenario"` default value to `true`.

<br />

```terraform
variable "enable-second-scenario" {
  type        = bool
  description = "Whether to enable the second scenario"
  default     = true
}
```

Next, issue the command `terraform apply` to deploy the second scenario. Notice how the `-var` flag is included with the token value in the command. 

<br />

```shell
terraform apply -var="token=931A3B02-8DCC-543F-A1B2-69423D1A0B94" -auto-approve
```

```
// Output condensed for readability
Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

> Log in to [Palette](https://console.spectrocloud.com) and navigate to the left **Main Menu**, click on **Apps**. Select the **scenario-2** row. Once you are on the scenario-2 overview page, click on the exposed URL for the service. A hyperlink for port 8080 and port 3000 is available.

Reviewer: ```suggestion
Log in to [Palette](https://console.spectrocloud.com) and navigate to the left **Main Menu**, click on **Apps**. Select the **scenario-2** row. When you are on the scenario-2 overview page, click on the exposed URL for the service. A hyperlink for port 8080 and port 3000 is available.
```

![A view of the scenario-2 overview page](/tutorials/deploy-app/devx_apps_deploy_scenario-2-overview.png)

Click on the UI’s service URL for port **8080** to access the Hello Universe application in a three-tier configuration.

<br />

<WarningBox>

  
It takes between one to three minutes for DNS to properly resolve the public load balancer URL. We recommend waiting a few moments before clicking on the service URL to prevent the browser from caching an unresolved DNS request.

</WarningBox>

![View of the self-hosted hello universe app](/tutorials/deploy-app/devx_apps_deploy-app_self-hosted-hello-universe.png)

> The global counter is no longer available; instead, you have a counter that starts at zero. Each time you click on the center image, the counter is incremented and stored in the Postgres database along with metadata.

Reviewer: ```suggestion
The global counter is no longer available. Instead, you have a counter that starts at zero. Each time you click on the center image, the counter is incremented and stored in the Postgres database along with metadata.
```

> ## Clean-up

Reviewer: ```suggestion
## Cleanup
```

Author: @ritawatson We have "clean-up" in the tutorial layout guide.

To remove all resources created in this tutorial, issue the `terraform destroy` command.

<br />

```shell
terraform destroy -var="token=931A3B02-8DCC-543F-A1B2-69423D1A0B94" -auto-approve
```

```
Destroy complete! Resources: 6 destroyed.
```

<br />

<InfoBox>

> If a cluster remains in the delete phase for over 15 minutes, it becomes eligible for **Force Delete**. To trigger a force delete, navigate to the respective cluster’s details page and click on **Settings**. Click on the **Force Delete Cluster** to delete the cluster. Palette will automatically remove clusters stuck in the cluster deletion phase for over 24 hrs.

Reviewer: ```suggestion
If a cluster remains in the delete phase for over 15 minutes, it becomes eligible for **Force Delete**. To trigger a force delete, navigate to the respective cluster’s details page and click on **Settings**. Click on the **Force Delete Cluster** to delete the cluster. Palette will automatically remove clusters stuck in the cluster deletion phase for over 24 hours.
```

</InfoBox>

<br />

</Tabs.TabPane>

</Tabs>

# Wrap-Up

In this tutorial, you learned about Palette’s Dev Engine and App Mode. You deployed two virtual clusters, each containing a different architecture and configuration of the Hello Universe application. Palette’s Dev Engine enables developers to quickly deploy applications into a Kubernetes environment without requiring Kubernetes knowledge. In a matter of minutes, you deployed a new Kubernetes cluster and all its applications without having to write Kubernetes configuration files.

> To learn more about Palette Dev Engine and its capabilities, check out the reference resource below.

Reviewer: ```suggestion
To learn more about Palette Dev Engine and its capabilities, check out the reference resources below.
```
